src/styles/evaluators.ts

```diff
@@ -0,0 +1,417 @@
+      T {
+    if (evaluatable instanceof Evaluator) {
+      return evaluatable.evaluate();
+    }
+
+    return evaluatable;
+  }
+
+  /**
+   * If the input is an Evaluator, returns the value of its isConstant property.
+   * Returns true for all other input values.
+   */
+  static isConstant<T>(evaluatable: Evaluatable<T>): boolean {
+    if (evaluatable instanceof Evaluator) {
+      return evaluatable.isConstant;
+    }
+    return true;
+  }
+
+  /**
+   * If true, the Evaluator will only evaluate its AST one time. If false, the
+   * Evaluator will re-evaluate the AST each time that the public evaluate
+   * method is invoked.
+   */
+  get isConstant(): boolean {
+    return false;
+  };
+
+  protected[$lastValue]: T|null = null;
+
+  /**
+   * This method must be implemented by subclasses. Its implementation should be
+   * the actual steps to evaluate the AST, and should return the evaluated
+   * result.
+   */
+  protected abstract[$evaluate](): T;
+
+  /**
+   * Evaluate the Evaluator and return the result. If the Evaluator is constant,
+   * the corresponding AST will only be evaluated once, and the result of
+   * evaluating it the first time will be returned on all subsequent
+   * evaluations.
+   */
+  evaluate(): T {
+    if (!this.isConstant || this[$lastValue] == null) {
+      this[$lastValue] = this[$evaluate]();
+    }
+    return this[$lastValue]!;
+  }
+}
+
+
+const $identNode = Symbol('identNode');
+
+/**
+ * Evaluator for CSS-like env() functions. Currently, only one environment
+ * variable is accepted as an argument for such functions: window-scroll-y.
+ *
+ * The env() Evaluator is explicitly dynamic because it always refers to
+ * external state that changes as the user scrolls, so it should always be
+ * re-evaluated to ensure we get the most recent value.
+ *
+ * Some important notes about this feature include:
+ *
+ *  - There is no such thing as a "window-scroll-y" CSS environment variable in
+ *    any stable browser at the time that this comment is being written.
+ *  - The actual CSS env() function accepts a second argument as a fallback for
+ *    the case that the specified first argument isn't set; our syntax does not
+ *    support this second argument.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/CSS/env
+ */
+export class EnvEvaluator extends Evaluator<NumberNode> {
+  protected[$identNode]: IdentNode|null = null;
+
+  constructor(envFunction: FunctionNode) {
+    super();
+
+    const identNode =
+        envFunction.arguments.length ? envFunction.arguments[0].terms[0] : null;
+
+    if (identNode != null && identNode.type === 'ident') {
+      this[$identNode] = identNode;
+    }
+  }
+
+  get isConstant(): boolean {
+    return false;
+  };
+
+  [$evaluate](): NumberNode {
+    if (this[$identNode] != null) {
+      switch (this[$identNode]!.value) {
+        case 'window-scroll-y':
+          const verticalScrollPosition = window.pageYOffset;
```

Comment:
Why do we want to have the calculation based on the absolute scroll position? It seems like it would be better UX to have it get updated in a relative way based on scrolling. It also seems like it would be simpler to accomplish this by intercepting a scroll event, but I assume there's a reason that doesn't work well?

Reply:
We do accomplish this by intercepting a `scroll` event. The `scroll` event is the signal that causes this code path to run (after the first time, that is). But, [the event object dispatched during scroll is a basic `Event` and doesn't give you any special information.](https://developer.mozilla.org/en-US/docs/Web/API/Document/scroll_event).

Reply:
This absolute behavior is growing on me. It is kind of nice to have this automatic viewport reset built in if someone gets lost zooming in on a model.